### DISCUSSION TITLE
onValidated reference event / tracking reference status

### DISCUSSION THREAD

**Discussion by ruudandriessen** [mobxjs/mobx-state-tree#1889] [Q&A] [answered]
Hi, in our project we've been heavily using MST & the references which they provide and enjoy them quite a bit 😄 

However, we found the need to track references in order to find models that are not referenced too - since these are 'unused' and we'd like to provide automated cleanups / warnings for our users. 

Since MST should already be tracking references, I thought this should conceptually be reasonably simple. We've tried some approaches, but we're not convinced on them:
* references only provide an `onInvalidated` event to hook into, no `onValidated` event. This would seem like the ideal solution for us.
*We tried using a custom reference with set/get. However, it didn't seem reliable, since these are triggered not on initial validation but every time whenever the reference has to be resolved.
* We could walk through the tree and count references ourselves, but this seems too expensive to do whenever a reference changes.
* We could define a computed that loops through all of the respective models and computes the used references. This seems like a feasible approach, but is still reasonably expensive & doesn't make use of references as effectively.

Any thoughts on adding an event like `onValidated` or ideas for an alternative approach?

**Comment by EmilTholin** [comment, marked as answer]
Hi @ruudandriessen!

I have not found a one-size-fits-all solution to this I'm 100% satisfied with myself, but I have opted for the 2nd alternative you outline when building a "Do you want to remove these unused objects?"-type user action.

It's all complicated further by MST lazily evaluating references (which is almost always desirable, but unfortunate in this case), so if all your references are not currently observed the other approaches you outline, or even a hypothetical `onValidated`, would most likely not be enough.

**Reply by ruudandriessen** [reply]
Yeah, that indeed validates my concerns. Thanks for providing some more details and at least confirming there's no easy way to use the existing references in MST.
